generator/src/main/java/com/graphicsfuzz/generator/transformation/DonateCodeTransformation.java

```diff
@@ -248,11 +248,22 @@ private Stmt prepareStatementToDonate(IInjectionPoint injectionPoint,
     final int maxTries = 10;
     int tries = 0;
     while (true) {
-      final Optional<TranslationUnit> maybeDonor = chooseDonor(generator);
+      final Optional<TranslationUnit> maybeDonor = chooseDonor(generator, shadingLanguageVersion);
+
       if (!maybeDonor.isPresent()) {
         // No compatible donors were found, thus we cannot do serious code donation here;
         // we return a null statement instead.
         return new NullStmt();
+      } else {
+        if (shadingLanguageVersion != maybeDonor.get().getShadingLanguageVersion()) {
+          throw new RuntimeException("Incompatible versions: ref=" + shadingLanguageVersion
+              + " donor=" + maybeDonor.get().getShadingLanguageVersion());
+        }
+        if (this.generationParams.getShaderKind() != maybeDonor.get().getShaderKind()) {
+          throw new RuntimeException("Incompatible shader types: ref="
```

Comment:
Please change "ref" to "reference" in the above.

Reply:
done